src/ucs/algorithm/crc.c

```diff
@@ -4,39 +4,66 @@
  * See file LICENSE for terms.
  */
 
-#include "crc.h"
+#include <ucs/algorithm/crc.h>
 
 #include <string.h>
 
-
-uint16_t ucs_crc16(const void *buffer, size_t size)
+/* This is CRC-16-CCITT with reversed (0x8408) polynomial representation */
+uint16_t ucs_crc16(uint16_t prev_crc, const void *buffer, size_t size)
 {
+    
+    const uint8_t *end = (const uint8_t*)(buffer + size);
+    uint16_t crc = ~prev_crc;
     const uint8_t *p;
-    uint16_t result;
-    uint8_t data;
-    int bit;
+    uint8_t data, bit;
 
     if (size == 0) {
-        return 0;
+        return crc;
     }
 
-    result = -1;
-    for (p = buffer; p < (const uint8_t*)(buffer + size); ++p) {
+    for (p = buffer; p < end; ++p) {
         data = *p;
         for (bit = 0; bit < 8; ++bit) {
-            result >>= 1;
-            if ((result ^ data) & 1) {
-                result = result ^ 0x8048;
+            crc >>= 1;
+            if ((crc ^ data) & 1) {
+                crc = crc ^ 0x8408;
             }
             data >>= 1;
         }
     };
 
-    result = ((result & 0xff) << 8) | ((result >> 8) & 0xff);
-    return ~result;
+    crc = ((crc & 0xff) << 8) | ((crc >> 8) & 0xff);
+    return ~crc;
 }
 
 uint16_t ucs_crc16_string(const char *s)
 {
-    return ucs_crc16((char*)s, strlen(s));
+    return ucs_crc16(0, (const char*)s, strlen(s));
+}
+
+/* This is CRC-32-IEEE with reversed (0xEDB88320) polynomial representation */
+uint32_t ucs_crc32(uint32_t prev_crc, const void *buffer, size_t size)
+{
+    const uint8_t *end = (const uint8_t*)(buffer + size);
+    uint32_t       crc = ~prev_crc;
+    const uint8_t *p;
+    uint8_t bit;
+
+    if (size == 0) {
+        return crc;
+    }
+
+    for (p = buffer; p < end; ++p) {
+        crc ^= *p;
+        for (bit = 0; bit < 8; bit++) {
+            crc = (crc >> 1) ^ (-(int)(crc & 1) & 0xedb88320);
+        }
+    }
+
+    return ~crc;
+}
+
+uint32_t ucs_crc32_string(const char *s)
```

Comment:
why are we exporting ucs_crc32_string(), i assume tcp would need to pass pointer and size?

Reply:
removed